File: solutions/cpp/allergies/allergies.cpp
#include "allergies.h"

namespace allergies
{
    using namespace std;

    string allergens[] = {
        "eggs",
        "peanuts",
        "shellfish",
        "strawberries",
        "tomatoes",
        "chocolate",
        "pollen",
        "cats",
    };

    Allergy_test allergy_test(int score)
    {
        unordered_set<string> allergies;
        int value = 1;
        for (const string &allergen : allergens)
        {
            if ((score & value) > 0)
            {
                allergies.insert(allergen);
            }
            value = value << 1;
        }
        return {allergies};
    }

    Allergy_test::Allergy_test(unordered_set<string> allergies)
        : allergies{allergies} {}

    bool Allergy_test::is_allergic_to(string allergen)
    {
        return allergies.find(allergen) != allergies.end();
    }

    unordered_set<string> Allergy_test::get_allergies()
    {
        return allergies;
    }
} // namespace allergies

File: solutions/lua/allergies/allergies.lua
local c_allergens = {
    eggs = 1,
    peanuts = 2,
    shellfish = 4,
    strawberries = 8,
    tomatoes = 16,
    chocolate = 32,
    pollen = 64,
    cats = 128
}

local function allergic_to(score, which)
    return score & c_allergens[which] > 0
end

local function ordered(allergens)
    table.sort(
        allergens,
        function(a, b)
            return c_allergens[a] < c_allergens[b]
        end
    )
    return allergens
end

local function list(score)
    local allergies = {}
    for allergen in pairs(c_allergens) do
        if allergic_to(score, allergen) then
            table.insert(allergies, allergen)
        end
    end
    return ordered(allergies)
end

return {list = list, allergic_to = allergic_to}
